src/libcore/stream.rs

```diff
@@ -0,0 +1,98 @@
+//! Asynchronous sequences of values.
+#![unstable(feature = "stream", issue = "none")]
+use crate::marker::Unpin;
+use crate::ops::DerefMut;
+use crate::pin::Pin;
+use crate::task::{Context, Poll};
+
+/// A stream of values produced asynchronously.
+///
+/// If `Future<Output = T>` is an asynchronous version of `T`, then `Stream<Item = T>` is an asynchronous version of
+/// `Iterator<Item = T>`. A stream represents a sequence of value-producing events that occur asynchronously to the
+/// caller.
+///
+/// The trait is modeled after `Future`, but allows `poll_next` to be called even after a value has been produced,
+/// yielding `None` once the stream has been fully exhausted.
+#[unstable(feature = "stream", issue = "none")]
+pub trait Stream {
+    /// Values yielded by the stream.
+    type Item;
+
+    /// Attempt to pull out the next value of this stream, registering the current task for wakeup if the value is not
+    /// yet available, and returning `None` if the stream is exhausted.
+    ///
+    /// # Return value
+    ///
+    /// There are several possible return values, each indicating a distinct stream state:
+    ///
+    /// * `Poll::Pending` means that this stream's next value is not ready yet. Implementations will ensure that the
+    ///     current task will be notified when the next value may be ready.
+    /// * `Poll::Ready(Some(val))` means that the stream has successfully produced a value, `val`, and may produce
+    ///     further values on subsequent `poll_next` calls.
+    /// * `Poll::Ready(None)` means that the stream has terminated, and `poll_next` should not be invoked again.
+    ///
+    /// # Panics
+    ///
+    /// Once a stream is finished, i.e. `Ready(None)` has been returned, further calls may result in a panic or other
+    /// "bad behavior".
```

Comment:
```suggestion
    /// Once a stream has finished (returned `Ready(None)` from `poll_next`), calling its
    /// `poll_next` method again may panic, block forever, or cause other kinds of
    /// problems; the `Stream` trait places no requirements on the effects of
    /// such a call. However, as the `poll_next` method is not marked `unsafe`,
    /// Rust's usual rules apply: calls must never cause undefined behavior
    /// (memory corruption, incorrect use of `unsafe` functions, or the like),
    /// regardless of the stream's state.
```

As with [`Future`'s documentation](https://docs.rs/futures/0.3.5/futures/prelude/trait.Future.html#panics), it needs to be mentioned that "bad behavior" does not contain UB.